File: javaclasses/RPGClass.java
package javaclasses;
import java.util.Map;
import java.util.List;

public class RPGClass {
  private String name;
  private Map<Integer, List<Feature>> classFeatures;
  private Map<Integer, List<Spell>> availableSpells;
  private List<Gear> startingInventory;

  // METHOD: basic constructor
  // initializes data structures
  public RPGClass() {
    // TODO
  }

  // METHOD: more in-depth constructor
  public RPGClass(String name, Map<Integer, List<Feature>> classFeatures, Map<Integer, List<Spell>> availableSpells, List<Gear> startingInventory) {
    // TODO
  }

  // TODO: methods such as 
  public List<Feature> featuresAt(int level) {
    return this.classFeatures.get(level);
  }

  public void /* THIS IS WRONG */ spellsAt(int level) {

  }

  /////////////// Java Object Methods /////////////////////////

  // METHOD: cloning method
  // returns a copy of the character
  @Override
  public RPGClass clone(){
    return this; 
    // TODO
  }

  // METHOD: determines if two characters are the same
  @Override
  public boolean equals(Object other){
    return false;
    // TODO
  }

  // METHOD: allows for characters to be sorted in a list
  @Override
  public int hashCode() {
    // TODO
    return name.hashCode();
  }

  @Override
  public String toString() {
    // TODO
    return "";
  }

  /////////////// Getters and Setters /////////////////////////
  // Getter and Setter methods for name field
  public String getName() {
    return name;
  }

  public void setName(String name) {
    this.name = name;
  }

  // Getter and Setter methods for classFeatures field
  public Map<Integer, List<Feature>> getClassFeatures() {
    return classFeatures;
  }

  public void setClassFeatures(Map<Integer, List<Feature>> classFeatures) {
    this.classFeatures = classFeatures;
  }

  // Getter and Setter methods for availableSpells field
  public Map<Integer, List<Spell>> getAvailableSpells() {
    return availableSpells;
  }

  public void setAvailableSpells(Map<Integer, List<Spell>> availableSpells) {
    this.availableSpells = availableSpells;
  }

  // Getter and Setter methods for startingInventory field
  public List<Gear> getStartingInventory() {
    return startingInventory;
  }

  public void setStartingInventory(List<Gear> startingInventory) {
    this.startingInventory = startingInventory;
  }
}

File: src/scripts/RPGClass.js
class RPGClass {
  #name
  #classFeatures
  #availableSpells
  #startingInventory

  constructor() {
    this.#name = ""
    this.#classFeatures = new Map()
    this.#availableSpells = new Map()
    this.#startingInventory = new Array()
  }

  constructor(name, classFeatures, availableSpells, startingInventory) {
    this.#name = name;
    this.#classFeatures = classFeatures;
    this.#availableSpells = availableSpells;
    this.#startingInventory = startingInventory;
  }

  // Getter and setter for name
  get name() { return this.#name; }
  set name(value) { this.#name = value; }

  // Getter and setter for classFeatures
  get classFeatures() { return this.#classFeatures; }
  getClassFeatures(level) { return this.#classFeatures.get(level) }
  set classFeatures(value) { this.#classFeatures = value; }
  addClassFeature(level, value) { this.#classFeatures.set(level, new Array(value)) }
  setLevelFeatures(level, arr) { this.#classFeatures.set(level, arr) }

  // Getter and setter for availableSpells
  get availableSpells() { return this.#availableSpells; }
  getAvailableSpells(level) { return this.#availableSpells.get(level) }
  set availableSpells(value) { this.#availableSpells = value; }
  addAvailableSpells(level, value) { this.#availableSpells.set(level, new Array(value)) }
  setAvailableSpells(level, arr) { this.#availableSpells.set(level, arr) }


  // Getter and setter for startingInventory
  get startingInventory() { return this.#startingInventory; }
  set startingInventory(value) { this.#startingInventory = value; }
  addStartingInventory(value) { this.#startingInventory.push(value) }
}

// Export the class
export default RPGClass;